File: retrocpu_test_framework/src/json_suite.ts
/**
 * テスト設定パスのプレースホルダ展開
 * 根拠: test_framework.mdc
 */

import fs from "node:fs";
import path from "node:path";
import { FRAMEWORK_BUILD, MONITOR_HEX, MONITOR_SRC, REPO_ROOT } from "./repo.js";

const PLACEHOLDERS: Record<string, string> = {
  MONITOR_SRC,
  MONITOR_HEX,
  FRAMEWORK_BUILD,
  REPO_ROOT,
};

/**
 * `${MONITOR_SRC}` などのプレースホルダを展開する。
 * @param text 文字列
 * @returns 展開後
 */
export function expandPlaceholders(text: string): string {
  return text.replace(/\$\{([A-Z_]+)\}/g, (all, key: string) => {
    const v = PLACEHOLDERS[key];
    if (v === undefined) {
      throw new Error(`Unknown placeholder ${all}`);
    }
    return v;
  });
}

/**
 * 設定内のパスを絶対パスにする（基準ディレクトリ → リポジトリルートの順）。
 * @param spec パス文字列
 * @param fromDir 相対パスの基準
 * @returns 絶対パス
 */
export function resolveSuitePath(spec: string, fromDir: string): string {
  const expanded = expandPlaceholders(spec);
  if (path.isAbsolute(expanded)) {
    return expanded;
  }
  const fromHere = path.resolve(fromDir, expanded);
  if (fs.existsSync(fromHere)) {
    return fromHere;
  }
  const fromRepo = path.resolve(REPO_ROOT, expanded);
  if (fs.existsSync(fromRepo)) {
    return fromRepo;
  }
  return fromHere;
}

File: retrocpu_test_framework_rs/src/json_suite.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};

use crate::error::FrameworkError;
use crate::repo::{framework_build, monitor_hex, monitor_src, repo_root};

fn placeholders() -> HashMap<&'static str, String> {
    HashMap::from([
        ("MONITOR_SRC", monitor_src().to_string_lossy().to_string()),
        ("MONITOR_HEX", monitor_hex().to_string_lossy().to_string()),
        (
            "FRAMEWORK_BUILD",
            framework_build().to_string_lossy().to_string(),
        ),
        ("REPO_ROOT", repo_root().to_string_lossy().to_string()),
    ])
}

pub fn expand_placeholders(text: &str) -> Result<String, FrameworkError> {
    let table = placeholders();
    let mut out = String::with_capacity(text.len());
    let mut i = 0;
    while let Some(start_rel) = text[i..].find("${") {
        let start = i + start_rel;
        out.push_str(&text[i..start]);

        let Some(end_rel) = text[start + 2..].find('}') else {
            return Err(FrameworkError::invalid_argument(
                "placeholder missing closing }".to_string(),
            ));
        };
        let end = start + 2 + end_rel;
        let key = &text[start + 2..end];

        let Some(value) = table.get(key) else {
            return Err(FrameworkError::invalid_argument(format!(
                "Unknown placeholder ${{{key}}}"
            )));
        };
        out.push_str(value);
        i = end + 1;
    }

    out.push_str(&text[i..]);
    Ok(out)
}

pub fn resolve_suite_path(spec: &str, from_dir: &Path) -> Result<PathBuf, FrameworkError> {
    let expanded = expand_placeholders(spec)?;
    let p = PathBuf::from(&expanded);
    if p.is_absolute() {
        return Ok(p);
    }

    let from_here = from_dir.join(&p);
    if from_here.exists() {
        return Ok(from_here);
    }

    let from_repo = repo_root().join(&p);
    if from_repo.exists() {
        return Ok(from_repo);
    }

    Ok(from_here)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn expands_known_placeholder() {
        let s = expand_placeholders("${REPO_ROOT}/abc").expect("expand should work");
        assert!(s.ends_with("/abc"));
    }

    #[test]
    fn rejects_unknown_placeholder() {
        let err = expand_placeholders("${UNKNOWN}").expect_err("unknown should fail");
        assert!(format!("{err}").contains("Unknown placeholder"));
    }

    #[test]
    fn resolve_prefers_from_dir_if_exists() {
        let mut tmp = std::env::temp_dir();
        tmp.push(format!(
            "tf-rs-json-suite-{}",
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap_or_default()
                .as_nanos()
        ));
        std::fs::create_dir_all(&tmp).expect("tmp create");
        let file = tmp.join("x.json");
        std::fs::write(&file, b"{}\n").expect("write");

        let resolved = resolve_suite_path("x.json", &tmp).expect("resolve");
        assert_eq!(resolved, file);

        let _ = std::fs::remove_file(file);
        let _ = std::fs::remove_dir_all(tmp);
    }
}
